sympy/functions/special/gamma_functions.py

```diff
@@ -307,6 +307,8 @@ def eval(cls, a, x):
             elif a.is_Integer or (2*a).is_Integer:
                 b = a - 1
                 if b.is_positive:
+                    if a.is_integer:
+                        return factorial(b) - exp(-x) * factorial(b) * sum([x ** k / factorial(k) for k in range(a)])
```

Comment:
Python's `sum` is inefficient for SymPy objects, see #13945. Since the point of PR is to optimize for large a, I recommend `Add(*[x ** k / factorial(k) for k in range(a)])`, and similarly for uppergamma